File: code/trees/bst-node.cpp
// A BST node (Sedgewick's ordered symbol table): key, value, two links, and a
// subtree-size count. get returns a pointer to the value or null; put inserts.
struct Node {
    Key   key;   Value val;
    Node* left  = nullptr;
    Node* right = nullptr;
    int   size  = 1;                    // # nodes in this subtree
};
int size(Node* t) { return t ? t->size : 0; }      // null-safe

Value* get(Node* t, const Key& key) {
    if (!t) return nullptr;             // miss: fell off a null link
    if (key == t->key) return &t->val;  // hit
    return key < t->key ? get(t->left, key) : get(t->right, key);
}
Node* put(Node* t, const Key& key, const Value& val) {
    if (!t) return new Node{key, val};  // grow at the null link
    if      (key < t->key) t->left  = put(t->left,  key, val);
    else if (key > t->key) t->right = put(t->right, key, val);
    else t->val = val;                  // equal key: update the value
    t->size = 1 + size(t->left) + size(t->right);   // recount on the way up
    return t;
}

File: code/trees/bst-node.rs
// A BST node with a subtree-size count. Links are Option<Box<Node>>, so the
// tree owns its children and drops them automatically (no manual free).
struct Node {
    key: i64, val: String,
    left: Option<Box<Node>>, right: Option<Box<Node>>,
    size: usize,
}
fn size(t: &Option<Box<Node>>) -> usize { t.as_ref().map_or(0, |n| n.size) }

fn get<'a>(t: &'a Option<Box<Node>>, key: i64) -> Option<&'a str> {
    match t {
        None => None,                               // miss: null link
        Some(n) if key == n.key => Some(&n.val),    // hit
        Some(n) if key <  n.key => get(&n.left, key),
        Some(n) => get(&n.right, key),
    }
}
fn put(t: Option<Box<Node>>, key: i64, val: String) -> Option<Box<Node>> {
    let mut n = match t {
        None => return Some(Box::new(               // grow at the null link
            Node { key, val, left: None, right: None, size: 1 })),
        Some(n) => n,
    };
    if      key < n.key { n.left  = put(n.left.take(),  key, val); }
    else if key > n.key { n.right = put(n.right.take(), key, val); }
    else { n.val = val; }                           // equal key: update
    n.size = 1 + size(&n.left) + size(&n.right);
    Some(n)
}
